### ISSUE TITLE
Unexpected behavior of allow_duplicates directive in Ansible 2.1.0

### ISSUE THREAD

**Issue by amitmodak** [ansible/ansible#16044]
##### ISSUE TYPE

- Bug Report
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = 
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

None

##### OS / ENVIRONMENT

N/A

##### SUMMARY

In a scenario where multiple roles depend on a single role, the dependent role is not executed multiple times in spite of setting `allow_duplicates: yes` in `meta/main.yml` of the dependent role. With Ansible 1.9.4, the dependent role does execute multiple times as expected.

Reference: https://github.com/ansible/ansible/issues/11707

##### STEPS TO REPRODUCE

I will try to explain the scenario with 2 roles `r1` and `r2` - both of which depend on a role `c`

- Playbook and Inventory

```
$ cat playbook.yml 

---
- name: dummy
  hosts: myhosts
  roles:
    - role: r1
      when: "inventory_hostname == 'host1'"
    - role: r2
      when: "inventory_hostname == 'host2'"

$ cat inv 
[myhosts]
host1 ansible_ssh_host=127.0.0.1       ansible_connection=local
host2 ansible_ssh_host=127.0.0.1       ansible_connection=local
```
- Role **r1**

```
$ cat roles/r1/tasks/main.yml 

---
- name: dummy1
  debug:
    msg: "I am r1"

$ cat roles/r1/meta/main.yml 

---
dependencies:
  - c
```
- Role **r2**

```
$ cat roles/r2/tasks/main.yml 

---
- name: dummy2
  debug:
    msg: "I am r2"

$ cat roles/r2/meta/main.yml 

---
dependencies:
  - c
```
- Role **c**

```
$ cat roles/c/tasks/main.yml 

---
- name: common
  debug:
    msg: "I am c"

$ cat roles/c/meta/main.yml 

---
allow_duplicates: yes
```

##### EXPECTED RESULTS

As `allow_duplicates: yes` has been configured, the execution order should have been c _(on host1)_ --> r1 _(on host1)_ --> c _(on host2)_ --> r2 _(on host2)_. Ansible 1.9.4 does have this execution order.
###### Output with Ansible 1.9.4

```
$ ansible-playbook -i inv playbook.yml 

PLAY [dummy] ****************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [host1]
ok: [host2]

TASK: [c | common] ************************************************************ 
skipping: [host2]
ok: [host1] => {
    "msg": "I am c"
}

TASK: [r1 | dummy1] *********************************************************** 
skipping: [host2]
ok: [host1] => {
    "msg": "I am r1"
}

TASK: [c | common] ************************************************************ 
skipping: [host1]
ok: [host2] => {
    "msg": "I am c"
}

TASK: [r2 | dummy2] *********************************************************** 
skipping: [host1]
ok: [host2] => {
    "msg": "I am r2"
}

PLAY RECAP ******************************************************************** 
host1                      : ok=3    changed=0    unreachable=0    failed=0   
host2                      : ok=3    changed=0    unreachable=0    failed=0   
```
##### ACTUAL RESULTS

c was not executed for the second time as a dependency of r2

```
$ ansible-playbook -i inv playbook.yml 

PLAY [dummy] *******************************************************************

TASK [setup] *******************************************************************
ok: [host2]
ok: [host1]

TASK [c : common] **************************************************************
ok: [host1] => {
    "msg": "I am c"
}
skipping: [host2]

TASK [r1 : dummy1] *************************************************************
skipping: [host2]
ok: [host1] => {
    "msg": "I am r1"
}

TASK [c : common] **************************************************************
skipping: [host1]
skipping: [host2]

TASK [r2 : dummy2] *************************************************************
skipping: [host1]
ok: [host2] => {
    "msg": "I am r2"
}

PLAY RECAP *********************************************************************
host1                      : ok=3    changed=0    unreachable=0    failed=0   
host2                      : ok=2    changed=0    unreachable=0    failed=0   
```

**Comment by ParagDoke** [comment]
+1

**Comment by kustodian** [comment]
I tested this a little bit and the problem is with the `when:` clause; without it everything works. It basically says the condition check failed when it checks it on a dependency the second time:

```
$ ansible-playbook -i inv test_role.yml -v 

PLAY [dummy] *******************************************************************

TASK [setup] *******************************************************************
ok: [host2]
ok: [host1]

TASK [c : debug] ***************************************************************
skipping: [host2] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
ok: [host1] => {
    "msg": "I am c"
}

TASK [r1 : debug] **************************************************************
skipping: [host2] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
ok: [host1] => {
    "msg": "I am r1"
}

TASK [c : debug] ***************************************************************
skipping: [host1] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
skipping: [host2] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}

TASK [r2 : debug] **************************************************************
skipping: [host1] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
ok: [host2] => {
    "msg": "I am r2"
}

PLAY RECAP *********************************************************************
host1                      : ok=3    changed=0    unreachable=0    failed=0   
host2                      : ok=2    changed=0    unreachable=0    failed=0
```

**Comment by kinghrothgar** [comment]
I am experiencing this problem without any conditionals in the mix.  My setup is as follows:
- `role A` has `role B` as a dependency twice with different variables
- `role B` has a dependency `role C`
- `role B` has `allow_duplicates: yes`

When I run the play, `role B` gets run twice as expected.  However, `role C` only gets run before the first `role B` run.

**Comment by phriscage** [comment]
+1

**Comment by jpcope** [comment]
I ran into this one today as well.
jimi-c commented on https://github.com/ansible/ansible/issues/11707 that this should be fixed now on devel branch

**Comment by jimi-c** [comment]
Confirming this is indeed fixed in devel, and additionally in `stable-2.1` as well (confirmed to work in v2.1.1.0-1):

<pre>
[jamesc@jimi 16044]$ ansible-playbook --version
ansible-playbook 2.1.2.0 (stable-2.1 c004ae578d) last updated 2016/09/20 11:31:23 (GMT -500)
  lib/ansible/modules/core: (detached HEAD 2b9300ec28) last updated 2016/09/20 15:08:33 (GMT -500)
  lib/ansible/modules/extras: (detached HEAD 35210ea450) last updated 2016/09/20 15:08:33 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/usr/share/ansible']
[jamesc@jimi 16044]$ ansible-playbook -vv test.yml -i inv 
Using /etc/ansible/ansible.cfg as config file
PLAYBOOK: test.yml *************************************************************
1 plays in test.yml
PLAY [dummy] *******************************************************************
TASK [setup] *******************************************************************
ok: [host2]
ok: [host1]
TASK [c : common] **************************************************************
task path: /home/jamesc/testing/16044/roles/c/tasks/main.yml:3
skipping: [host2] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
ok: [host1] => {
    "msg": "I am c"
}
TASK [r1 : dummy1] *************************************************************
task path: /home/jamesc/testing/16044/roles/r1/tasks/main.yml:3
skipping: [host2] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
ok: [host1] => {
    "msg": "I am r1"
}
TASK [c : common] **************************************************************
task path: /home/jamesc/testing/16044/roles/c/tasks/main.yml:3
skipping: [host1] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
ok: [host2] => {
    "msg": "I am c"
}
TASK [r2 : dummy2] *************************************************************
task path: /home/jamesc/testing/16044/roles/r2/tasks/main.yml:3
skipping: [host1] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
ok: [host2] => {
    "msg": "I am r2"
}
PLAY RECAP *********************************************************************
host1                      : ok=3    changed=0    unreachable=0    failed=0   
host2                      : ok=3    changed=0    unreachable=0    failed=0   
</pre>

As such, we will close it now. If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
